src/kOS.Safe/Execution/Stack.cs

```diff
@@ -41,6 +41,59 @@ public class Stack : IStack
 
         private int triggerContextCount = 0;
 
+        /// <summary>
+        /// Current size of argument stack
+        /// </summary>
+        public int ArgumentCount => argumentCount;
+        /// <summary>
+        /// Current size of scope stack
+        /// </summary>
+        public int ScopeCount => scopeCount;
+        /// <summary>
+        /// All arguments on the stack (from top to bottom)
+        /// </summary>
+        public IEnumerable<object> Arguments
+        {
+            get
+            {
+                for (int i = argumentCount; i > 0;)
+                    yield return argumentStack[--i];
```

Comment:
I never liked ``yield return``.  It requires too much magic collusion between the compiler (``yield`` being a keyword) and .NET (``System.Collections.Generic.IEnumerable``).

Reply:
Also, @firda-cze , I looked through this PR's full list of changed files and I'm not seeing where you use these new ``IEnumerable``s that you added to ``Stack.cs``.  You said you had them in here because you needed to use them elsewhere but where are you using them?

Reply:
Well, you can create full blown enumerator, but this is just easier. You write the main loop and `yield return` elements that you can consume through `foreach` later, a bit like async, but without a thread. Co-routines in unity are implemented the same way and they actually use it in thread-like (fiber) fashion.

See https://github.com/firda-cze/KOS/blob/firda/src/kOS.Tester/TesterCPU.cs for the reason I added these. Definitelly using `Arguments`, wanted to use `ArgumentsFromBottom` for checking what has changed, before I understood the argument order reversing in call. I did not create view for `Scopes` yet, but will need that too.

Here #2247 are pictures from the tester.

Reply:
Well, I disagree with it being easier, but that's a matter of opinion.  (In trying to find the sweet spot between being hard because it's too terse versus being hard because it's too verbose, my opinion is that the C# culture errs on the side of being too terse.  The C# community loves syntactic sugar a lot more than I do.)

I understand how much of a PITA it is to keep things separate in one PR that you're using in another PR when neither is merged yet.  @hvacengi is dealing with this particular merge so I'll defer to him to decide how to handle it.

By the way, is that tester something you plan to release?  Is it an independent executable or something that you invoke inside KSP's process?

Reply:
It is independent exe, only referencing the Safe.dll, feel free to take it any time, open-source, free to use, but it is not yet ready, it lacks the scopes.
But it can already step individual instructions, show opcodes, all jumps/calls, stack, output, log, all the tests and helped me a lot in understanding what is going on under the hood.
Basic WinForms, so, should be portable to Linux and Mac as well. Want [Eto](https://github.com/picoe/Eto) or Xamarin.Forms instead?

Reply:
I'm not well versed enough in C# development (I only learned C# specifically for kOS.  I come from the Java/C++ world on Unix mostly) to answer that question.  I don't really know the difference between the GUI options you mentioned.

Reply:
I have heard that one of the major speedups they did in one of the KSP versions was that they went through and replaced many of their `foreach` iterator-based loops with traditional old-school index-counting `for` loops because Mono's iterator implementation was much slower than just getting items by index (A problem specific to Mono's implementation that Unity uses to be cross-platform, not Microsoft's .NET).  That's part of the reason I never had much incentive to want to make an iterator-based wrapper around the stack access.

Reply:
Nothing against arrays and indexes, but that would require merging the stack into the CPU, so that derived classes (TesterCPU) can access it. Should not be exposed publicly, that is why I added this read-only interface for it.

Reply:
Foreach was fixed in Unity 5.5:
https://unity3d.com/learn/tutorials/topics/performance-optimization/optimizing-garbage-collection-unity-games

Reply:
To be honest, I never really understood why the heck in Marianoapp's original implementation, the stack has private protections against the CPU.  All that did is make me have to add a ton of dependency inversion interfaces between the two, when really the CPU class should have full permissions to do whatever it likes to the stack.  The only real reason I can see to protect the stack would be to deliberately bottleneck the access through a few key APIs, but those APIs really should allow all access and all major operations.